crates/http/src/backend/h3/mod.rs

```diff
@@ -0,0 +1,182 @@
+
+        let endpoint = h3_quinn::quinn::Endpoint::server(server_config, self.bind)?;
+
+        let tasks = (0..self.worker_tasks).map(|n| {
+            let ctx = self.ctx.clone();
+            let service_factory = service_factory.clone();
+            let endpoint = endpoint.clone();
+
+            let worker_fut = async move {
+                #[cfg(feature = "tracing")]
+                tracing::trace!("waiting for connections");
+
+                while let Some(Some(new_conn)) = endpoint.accept().with_context(&ctx).await {
+                    let mut service_factory = service_factory.clone();
+                    let ctx = ctx.clone();
+
+                    tokio::spawn(async move {
+                        let _res: Result<_, Error<F>> = async move {
+                            let conn = new_conn.await?;
+                            let addr = conn.remote_address();
+
+                            #[cfg(feature = "tracing")]
+                            tracing::debug!(addr = %addr, "accepted quic connection");
+
+                            let connection_fut = async move {
+                                let mut h3_conn = h3::server::Connection::new(h3_quinn::Connection::new(conn)).await?;
+
+                                // make a new service for this connection
+                                let http_service = service_factory
+                                    .new_service(addr)
+                                    .await
+                                    .map_err(|e| Error::ServiceFactoryError(e))?;
+
+                                loop {
+                                    match h3_conn.accept().with_context(&ctx).await {
+                                        Some(Ok(Some((req, stream)))) => {
+                                            #[cfg(feature = "tracing")]
+                                            tracing::debug!(method = %req.method(), uri = %req.uri(), "received request");
+
+                                            let (mut send, recv) = stream.split();
+
+                                            let size_hint = req
+                                                .headers()
+                                                .get(http::header::CONTENT_LENGTH)
+                                                .and_then(|len| len.to_str().ok().and_then(|x| x.parse().ok()));
+                                            let body = QuicIncomingBody::new(recv, size_hint);
+                                            let req = req.map(|_| crate::body::IncomingBody::from(body));
+
+                                            let ctx = ctx.clone();
+                                            let mut http_service = http_service.clone();
+                                            tokio::spawn(async move {
+                                                // The context must live as long as the request is being handled
+                                                let _ctx = ctx;
```

Comment:
I would replace this in favour of a `drop(ctx)` at the very end of this function.